CoreScriptsRoot/Modules/VR/VirtualKeyboard.lua

```diff
@@ -0,0 +1,678 @@
+-- VirtualKeyboard.lua --
+-- Written by Kip Turner, copyright ROBLOX 2016 --
+
+
+local CoreGui = game:GetService('CoreGui')
+local RunService = game:GetService('RunService')
+local UserInputService = game:GetService('UserInputService')
+local GuiService = game:GetService('GuiService')
+local HttpService = game:GetService('HttpService')
+local ContextActionService = game:GetService('ContextActionService')
+local PlayersService = game:GetService('Players')
+
+local RobloxGui = CoreGui:WaitForChild("RobloxGui")
+local Util = require(RobloxGui.Modules.Settings.Utility)
+
+local vrKeyboardSuccess, vrKeyboardFlagValue = pcall(function() return settings():GetFFlag("UseVRKeyboardInLua") end)
+local useVRKeyboard = (vrKeyboardSuccess and vrKeyboardFlagValue == true)
+
+
+local RENDER_STEP_NAME = game:GetService("HttpService"):GenerateGUID() .. "VirtualKeyboard"
+
+local NORMAL_KEY_COLOR = Color3.new(1,1,1)
+local HOVER_KEY_COLOR = Color3.new(178/255,178/255,178/255)
+local PRESSED_KEY_COLOR = Color3.new(0,162/255,1)
+local SET_KEY_COLOR = Color3.new(0,162/255,1)
+---------------------------------------- KEYBOARD LAYOUT --------------------------------------
+local KEYBOARD_LAYOUT = HttpService:JSONDecode([==[
+[
+  [
+    "~\n`",
+    "!\n1",
+    "@\n2",
+    "#\n3",
+    "$\n4",
+    "%\n5",
+    "^\n6",
+    "&\n7",
+    "*\n8",
+    "(\n9",
+    ")\n0",
+    "_\n-",
+    "+\n=",
+    {
+      "w": 2
+    },
+    "Delete"
+  ],
+  [
+    {
+      "w": 1.5
+    },
+    "Tab",
+    "Q",
+    "W",
+    "E",
+    "R",
+    "T",
+    "Y",
+    "U",
+    "I",
+    "O",
+    "P",
+    "{\n[",
+    "}\n]",
+    {
+      "w": 1.5
+    },
+    "|\n\\"
+  ],
+  [
+    {
+      "w": 1.75
+    },
+    "Caps",
+    "A",
+    "S",
+    "D",
+    "F",
+    "G",
+    "H",
+    "J",
+    "K",
+    "L",
+    ":\n;",
+    "\"\n'",
+    {
+      "w": 2.25
+    },
+    "Enter"
+  ],
+  [
+    {
+      "w": 2.25
+    },
+    "Shift",
+    "Z",
+    "X",
+    "C",
+    "V",
+    "B",
+    "N",
+    "M",
+    "<\n,",
+    ">\n.",
+    "?\n/",
+    {
+      "w": 2.75
+    },
+    "Shift"
+  ],
+  [
+    {
+      "x": 3.75,
+      "a": 7,
+      "w": 6.25
+    },
+    ""
+  ]
+]
+]==])
+---------------------------------------- END KEYBOARD LAYOUT --------------------------------------
+
+
+local function tokenizeString(str, tokenChar)
+	local words = {}
+	for word in string.gmatch(str, '([^' .. tokenChar .. ']+)') do
+	    table.insert(words, word)
+	end
+	return words
+end
+
+-- RayPlaneIntersection
+
+-- http://www.siggraph.org/education/materials/HyperGraph/raytrace/rayplane_intersection.htm
+local function RayPlaneIntersection(ray, planeNormal, pointOnPlane)
```

Comment:
this is unfortunate that we have to do this in lua, I guess the camera doesn't have a function that does this does it?